### ISSUE TITLE
Strip unittest dependency to fully upgrade tests to pytest, so we can provide better quality and coverage of tests

### ISSUE THREAD

**Issue by openSourcerer9000** [ioos/compliance-checker#874]
As things are currently, many of our tests are trying to test multiple cases per test, or don't fail gracefully, providing enough info to diagnose the problem. I believe we can give them a makeover if we can upgrade them to the pytest framework, which will increase our confidence in the codebase and new PR's that pass all tests. Pytest has all the features built in that can easily handle a lot of these common testing blues.

Since we are using pytest, we can run the legacy unittests, but they can't leverage pytest features like fixtures or parameterization, as they are all inheriting from the incompatible unittest TestCase object. 
From a discussion with @benjwadams:

https://github.com/benjwadams/compliance-checker/pull/4
```
[10:11 AM] Sean Micek    
    Ben Adams I noticed all of our test objects inherit from unittest TestCase but we have pytest now, is there any reason we couldn't upgrade them at some point so we can use pytest fixtures/parameterization?
Edited​[10:13 AM] Ben Adams
    Given that TestCase works with pytest, there's no reason why we couldn't change over parts of the library
​[10:14 AM] Sean Micek
    How I understand it, pytest made it backwards compatible, so if you want to switch it will still run all your old unittests. I know nothing about unittest, but when I tried adding pytest params I got an error complaining that the test was a unittest TestCase
​[10:16 AM] Sean Micek
    I guess it all stems from tests/_.init_.py
I would think we could just drop the unittest inheritance and it should work, pytest should run any method in the tests files with test_ prefix
```
https://github.com/ioos/compliance-checker/blob/e14be12c130d1100c6a6a7d700c539106049dc5e/compliance_checker/tests/__init__.py#L1-L9

I tried simply removing the TestCase inheritance from the BaseTestCase object, and got errors as all these unittest setUp and tearDown helper functions stopped working without the dependency. We can convert them all to pytest fixtures, but beyond that I'm not sure what other features we may be using from unittest, I'm just not familiar with the library. 

This probably isn't an urgent upgrade, but I feel that it would allow us to diagnose and patch issues more quickly and merge PR's more confidently.

![image](https://user-images.githubusercontent.com/61931826/127524607-8d434674-9039-49c6-b093-0c6f2e71b4d8.png)
![image](https://user-images.githubusercontent.com/61931826/127527977-d5dbb4ba-b7c8-4293-a38c-f051f2df5b24.png)

**Comment by benjwadams** [comment]
@ocefpaf, any thoughts on switching over to py.test?  I used fixtures in a project recently and found them quite useful.

**Comment by ocefpaf** [comment]
> @ocefpaf, any thoughts on switching over to py.test? I used fixtures in a project recently and found them quite useful.

Just one. YES! PLEASE!! ;-P

**Comment by ChrisBarker-NOAA** [comment]
Picking this up WAY later -- pytest is BOTH a test runner and a test framework.

So you can mix unittest and pytest tests in one project, even one file.

Anyway, that means that we can transition the tests to pytest tests slowly -- when you are messing around with a test, update it to pytest.

But no need to remove unittest if you're not doing anything else with a test.

**Comment by benjwadams** [comment]
Yeah, I'm in favor of the above approach, to gradually move over tests as needed.  Closing in favor of using both approaches as it's not really feasible to migrate all of the tests over to pytest idioms right now.  If anyone's interested, they may feel free to do so.

**Comment by ocefpaf** [comment]
While this is laborious, it is not too complex and would make for a nice GSoC 2025 small project.

**Comment by benjwadams** [comment]
Implemented by #1160, closing.
